py: Faster qstr search

amirgon opened the pull request:
Motivation: qstr search scales poorly - see https://forum.micropython.org/viewtopic.php?f=2&t=9748

In short, when there are many qstrs, performance is poor because of the way they are searched.
This affects, for example, load time of mpy modules since strings are replaced by qstrs during import.  
On ESP32 it takes many seconds loading few mpy files in the presence of large native libraries ([LVGL](https://github.com/lvgl/lvgl) for example).

---

Today qstr implementation scans strings sequntially.
In cases there are many strings this can become very inefficient.
This change improves qstr search performance by using binary search in
sorted qstr pools, when possible.

This change introduces an option to create a sorted string pool, which
is then searched by a binary search instead of sequential search.

qstr pool can be either "sorted" or "unsorted", whereas the unsorted is
searched sequentally as today.
Native modules (MP_ROM_QSTR) and frozen modules generate sorted pools.
Currently runtime strings are unsorted.

The constant string pools is split into two and a new pool is introduced,
"special_const_pool". This is required because the first sequence of
strings already requires special ordering therefore created unsorted,
while the rest of the constants are generated sorted.

qstr_find_strn searches strings in each pool. If the pool is sorted and
larger than a threshold, it will be search using binary search instead
of sequential search, significantly improving performance.

amirgon committed:
py: Faster qstr search.

Today qstr implementation scans strings sequntially.
In cases there are many strings this can become very inefficient.
This change improves qstr search performance by using binary search in
sorted qstr pools, when possible.

This change introduces an option to create a sorted string pool, which
is then searched by a binary search instead of sequential search.

qstr pool can be either "sorted" or "unsorted", whereas the unsorted is
searched sequentally as today.
Native modules (MP_ROM_QSTR) and frozen modules generate sorted pools.
Currently runtime strings are unsorted.

The constant string pools is split into two and a new pool is introduced,
"special_const_pool". This is required because the first sequence of
strings already requires special ordering therefore created unsorted,
while the rest of the constants are generated sorted.

qstr_find_strn searches strings in each pool. If the pool is sorted and
larger than a threshold, it will be search using binary search instead
of sequential search, significantly improving performance.

Signed-off-by: Amir Gonnen <amirgonnen@gmail.com>

stinos commented:
Do you have some measurements, preferably from different ports, showing how significant the improvement is?

amirgon commented:
> Do you have some measurements, preferably from different ports, showing how significant the improvement is in different situations?

Unfortunately I can only provide measurements for ESP32.   
On ESP32 @240Mhz I'm importing a 10KB view.mpy file shortly after reset:
```
>>> ticks_ms();import view;ticks_ms()
```
- Without this change the import completes in 3590ms
- With this change the import completes in 169ms 

That's x 21 performance gain.
Measuring this multiple times I'm getting the same results with up to 1 ms difference.

The firmware is Micropython 1.14 with LVGL native module.  
LVGL is a large library. Looking at `qstrdefs.generated.h` there are 4450 QDEFs. It's clear that scanning all these strings sequentially for each and every str-->qstr conversion is inefficient, particularly since the const qstr pool is stored on flash.

> And if the search is faster what is the effect on creating and adding new qstr values at runtime?

There is no effect on creating and adding new qstr values at runtime because runtime qstr pools remain unsorted, and scanned in the same manner as today.
Only qstr pools generated at compile time are sorted, so the benefit is for native and frozen modules only.

stinos commented:
>  runtime qstr pools remain unsorted

Oops sorry missed that. Any particular reason for this?

amirgon commented:
> > runtime qstr pools remain unsorted
> 
> Oops sorry missed that. Any particular reason for this?

Yes.

- qstrs are indices into qstr pools. So once a pool is in use (the code refers to specific qstrs in the pool), we can no longer change its sort order otherwise existing qstrs used in the code would no longer be correct.  
To overcome this we would either need to add an indirection level when accessing qstrs (that would cost performance and RAM) or select a different solution such as hash-table instead of sorted array. Hash-table is a valid alternative, but has its disadvantages (more costly in RAM, harder to tune optimally, more complex to implement).
- From practical point of view, when scaling up I think we can expect the mass of qstrs to come from native or frozen modules, and the minority to be runtime qstrs. So most benefit would be achieved from improving access to native / frozen qstrs.

stinos commented:
Thanks for the explanation, makes sense. You might want to fix code formatting, some grammar issues in comments.

amirgon commented:
> You might want to fix code formatting, some grammar issues in comments.

Done.

Gadgetoid commented:
Just to weigh in here, I've been shipping a MicroPython build for PicoSystem that includes this patch since October 2021 and - so far - have had no problems with it. The speedup (alongside doing some heinous things with builtins) is essential for our wide/flat API surface.

thyrrestrup commented:
Regarding indirection when accessing qstr’s see #8191

dpgeorge commented:
Running some benchmarks on this PR, it looks very good.  On PYBv1.1 which has about 1600 static qstrs:
```
diff of scores (higher is better)
N=100 M=100                  pyb-p0 ->     pyb-p1         diff      diff% (error%)
bm_chaos.py                  369.56 ->     374.23 :      +4.67 =  +1.264% (+/-0.02%)
bm_fannkuch.py                77.51 ->      78.45 :      +0.94 =  +1.213% (+/-0.00%)
bm_fft.py                   2558.76 ->    2577.15 :     +18.39 =  +0.719% (+/-0.00%)
bm_float.py                 5922.05 ->    6143.30 :    +221.25 =  +3.736% (+/-0.02%)
bm_hexiom.py                  47.96 ->      49.77 :      +1.81 =  +3.774% (+/-0.00%)
bm_nqueens.py               4446.35 ->    4457.18 :     +10.83 =  +0.244% (+/-0.00%)
bm_pidigits.py               647.41 ->     650.75 :      +3.34 =  +0.516% (+/-0.15%)
core_qstr.py                5020.31 ->   16570.01 :  +11549.70 = +230.059% (+/-0.01%)
misc_aes.py                  421.88 ->     427.24 :      +5.36 =  +1.271% (+/-0.01%)
misc_mandel.py              3566.58 ->    3584.88 :     +18.30 =  +0.513% (+/-0.00%)
misc_pystone.py             2487.97 ->    2583.21 :     +95.24 =  +3.828% (+/-0.02%)
misc_raytrace.py             376.64 ->     383.84 :      +7.20 =  +1.912% (+/-0.02%)
viper_call0.py               576.54 ->     576.73 :      +0.19 =  +0.033% (+/-0.04%)
viper_call1a.py              550.05 ->     550.38 :      +0.33 =  +0.060% (+/-0.04%)
viper_call1b.py              435.39 ->     435.95 :      +0.56 =  +0.129% (+/-0.19%)
viper_call1c.py              441.49 ->     440.80 :      -0.69 =  -0.156% (+/-0.21%)
viper_call2a.py              535.74 ->     536.31 :      +0.57 =  +0.106% (+/-0.12%)
viper_call2b.py              381.40 ->     381.91 :      +0.51 =  +0.134% (+/-0.15%)
```
The new test there, `core_qstr.py`, effectively calls `qstr_find_strn()` many times in a loop.  And it's significantly faster with this PR.  Other benchmarks are not impacted, which is good.

On unix (which is always noisy) it's possible to also see an improvement, although unix only has about 600 static qstrs:
```
diff of scores (higher is better)
N=5000 M=100                unix-p0 ->    unix-p1         diff      diff% (error%)
bm_fannkuch.py                10.52 ->      11.00 :      +0.48 =  +4.563% (+/-4.87%)
bm_fft.py                 221486.06 ->  222913.26 :   +1427.20 =  +0.644% (+/-5.24%)
bm_float.py               446655.34 ->  452957.42 :   +6302.08 =  +1.411% (+/-4.16%)
bm_hexiom.py                3531.20 ->    3549.03 :     +17.83 =  +0.505% (+/-2.11%)
bm_nqueens.py             442323.78 ->  414218.36 :  -28105.42 =  -6.354% (+/-8.21%)
bm_pidigits.py             48953.40 ->   50719.93 :   +1766.53 =  +3.609% (+/-7.57%)
core_qstr.py              940349.56 -> 1432698.78 : +492349.22 = +52.358% (+/-7.42%)
misc_aes.py                33653.40 ->   32860.59 :    -792.81 =  -2.356% (+/-7.40%)
misc_mandel.py            242690.30 ->  238764.20 :   -3926.10 =  -1.618% (+/-6.17%)
misc_pystone.py           139726.47 ->  135769.40 :   -3957.07 =  -2.832% (+/-7.61%)
misc_raytrace.py           37689.06 ->   36494.55 :   -1194.51 =  -3.169% (+/-7.47%)
viper_call0.py             34051.60 ->   34323.85 :    +272.25 =  +0.800% (+/-6.95%)
viper_call1a.py            31946.89 ->   33872.77 :   +1925.88 =  +6.028% (+/-7.69%)
viper_call1b.py            27876.34 ->   26771.29 :   -1105.05 =  -3.964% (+/-6.14%)
viper_call1c.py            26162.97 ->   26561.40 :    +398.43 =  +1.523% (+/-10.81%)
viper_call2a.py            27179.72 ->   33407.50 :   +6227.78 = +22.913% (+/-12.60%)
viper_call2b.py            19137.07 ->   24199.32 :   +5062.25 = +26.453% (+/-8.79%)
```

dpgeorge commented on py/qstr.c:
```diff
@@ -171,14 +194,48 @@ STATIC qstr qstr_add(const byte *q_ptr) {
+
+        // sequential search for the remaining strings
+        for (const byte **q = pool->qstrs + low; q != pool->qstrs + high + 1; q++) {
+            if (*q &&
```
@amirgon why does it check for `*q` being non-zero?  The original code did not need to do this in the linear search part.

amirgon replied:
I removed this check after rebase.

dpgeorge commented on py/qstr.c:
```diff
@@ -171,14 +194,48 @@ STATIC qstr qstr_add(const byte *q_ptr) {
     return MP_STATE_VM(last_pool)->total_prev_len + MP_STATE_VM(last_pool)->len - 1;
 }
 
+#define MP_QSTR_SEARCH_THRESHOLD 10
```
How was this number chosen?

amirgon replied:
The rational was that sequential search over small sequences is more efficient than binary search because of caching, compiler optimizations etc. It's probably a different number for different architectures, compiler flags etc.  

So this number was chosen simply because it was good enough on my tests.

dpgeorge commented on py/qstr.c:
```diff
@@ -171,14 +194,48 @@ STATIC qstr qstr_add(const byte *q_ptr) {
+                if (len > str_len) {
+                    len = str_len;
+                }
+                int cmp = memcmp(Q_GET_DATA(*q), str, str_len);
```
It might work to sort on qstr hash (and length) instead of qstr data, and then this comparison here would just be a comparison (subtraction) of the 2 hashes, which would be a lot faster than a memcmp.

Once it reached a case of 2 identical hashes it would then need to use memcmp, or just fall through to a linear search.

amirgon replied:
This is a good idea so I gave it a shot.  

It turned out to be more challenging than I expected:
As you can see, [extra_coverage](https://github.com/micropython/micropython/runs/5325683711?check_suite_focus=true) test fails. That's because it tests auto completions which prints qstrs in order(`find_completions`, `print_completions`). So once qstrs are sorted by hash, the autocompletion order is no longer alphabetical...

We have a few options:
- Sort the auto completions results (allocate ram and sort qstr range by name?)
- Keep autocompletions unsorted
- Give up sorting by hash and keep alphabetical qstr sort

What do you think?

dpgeorge replied:
Would be good to run some benchmarks on sorting by hash vs string content.  Do you have the original version which used string sorting?

amirgon replied:
> Do you have the original version which used string sorting?

Yes. Only the last commit on this PR introduced hash sorting.  
Simply checkout previous commmit 5808635e4

amirgon committed:
py/makeqstrdata.py: Refactor qstr class.

Use Qstr class instead of tuple, where properties are calculated only when
accessed.
This is needed as preparation to using hash as the sort key instead the
qstr string. It also makes the code more readable when referring to a qstr
in py/makeqstrdata.py and tools/mpy-tool.py (for example, refer to q.order
instead of q[0], or q.qstr instead of q[2])

Signed-off-by: Amir Gonnen <amirgonnen@gmail.com>

amirgon committed:
py/qstr: Sort qstrs by hash.

Instead of sorting by string, sort qstrs by (hash, len).
This allows faster binary serach on qstr_find_strn, since it's faster
to compare hashes than strings.
A few strings needed to be moved to special string pool (QDEF0)
because their qstr is assumed to be small (8 bit) on py/scope.c

Signed-off-by: Amir Gonnen <amirgonnen@gmail.com>

dpgeorge commented:
I did performance tests for sorting on strings and sorting on hashes (on PYBv1.0).

Going from baseline (current master) to sorting them on string value gives:
```
diff of scores (higher is better)
N=100 M=100                pyb-baseline -> pyb-sort-str         diff      diff% (error%)
bm_chaos.py                    357.40 ->     352.80 :      -4.60 =  -1.287% (+/-0.02%)
bm_fannkuch.py                  77.49 ->      77.40 :      -0.09 =  -0.116% (+/-0.01%)
bm_fft.py                     2539.26 ->    2525.13 :     -14.13 =  -0.556% (+/-0.00%)
bm_float.py                   5908.33 ->    5872.71 :     -35.62 =  -0.603% (+/-0.01%)
bm_hexiom.py                    47.93 ->      48.77 :      +0.84 =  +1.753% (+/-0.00%)
bm_nqueens.py                 4459.94 ->    4433.80 :     -26.14 =  -0.586% (+/-0.00%)
bm_pidigits.py                 644.88 ->     640.95 :      -3.93 =  -0.609% (+/-0.27%)
core_import_mpy_multi.py       581.49 ->     759.37 :    +177.88 = +30.590% (+/-0.00%)
core_import_mpy_single.py       67.16 ->     105.41 :     +38.25 = +56.954% (+/-0.01%)
core_qstr.py                    64.12 ->     206.44 :    +142.32 = +221.959% (+/-0.01%)
core_yield_from.py             354.50 ->     353.96 :      -0.54 =  -0.152% (+/-0.00%)
misc_aes.py                    405.59 ->     410.03 :      +4.44 =  +1.095% (+/-0.00%)
misc_mandel.py                3416.48 ->    3387.86 :     -28.62 =  -0.838% (+/-0.00%)
misc_pystone.py               2405.57 ->    2419.06 :     +13.49 =  +0.561% (+/-0.00%)
misc_raytrace.py               374.01 ->     368.50 :      -5.51 =  -1.473% (+/-0.00%)
```
Going from baseline to sorting on hash:
```
diff of scores (higher is better)
N=100 M=100                pyb-baseline -> pyb-sort-hash         diff      diff% (error%)
bm_chaos.py                    357.40 ->     359.03 :      +1.63 =  +0.456% (+/-0.03%)
bm_fannkuch.py                  77.49 ->      76.28 :      -1.21 =  -1.561% (+/-0.01%)
bm_fft.py                     2539.26 ->    2523.73 :     -15.53 =  -0.612% (+/-0.00%)
bm_float.py                   5908.33 ->    5910.11 :      +1.78 =  +0.030% (+/-0.01%)
bm_hexiom.py                    47.93 ->      48.85 :      +0.92 =  +1.919% (+/-0.00%)
bm_nqueens.py                 4459.94 ->    4421.59 :     -38.35 =  -0.860% (+/-0.00%)
bm_pidigits.py                 644.88 ->     644.49 :      -0.39 =  -0.060% (+/-0.21%)
core_import_mpy_multi.py       581.49 ->     763.36 :    +181.87 = +31.277% (+/-0.00%)
core_import_mpy_single.py       67.16 ->     106.25 :     +39.09 = +58.204% (+/-0.01%)
core_qstr.py                    64.12 ->     212.31 :    +148.19 = +231.114% (+/-0.00%)
core_yield_from.py             354.50 ->     354.05 :      -0.45 =  -0.127% (+/-0.00%)
misc_aes.py                    405.59 ->     404.15 :      -1.44 =  -0.355% (+/-0.00%)
misc_mandel.py                3416.48 ->    3397.47 :     -19.01 =  -0.556% (+/-0.00%)
misc_pystone.py               2405.57 ->    2446.53 :     +40.96 =  +1.703% (+/-0.00%)
misc_raytrace.py               374.01 ->     369.21 :      -4.80 =  -1.283% (+/-0.00%)
```
Difference going from sorting on string to sorting on hash is small:
```
diff of scores (higher is better)
N=100 M=100                pyb-sort-str -> pyb-sort-hash         diff      diff% (error%)
bm_chaos.py                    352.80 ->     359.03 :      +6.23 =  +1.766% (+/-0.01%)
bm_fannkuch.py                  77.40 ->      76.28 :      -1.12 =  -1.447% (+/-0.00%)
bm_fft.py                     2525.13 ->    2523.73 :      -1.40 =  -0.055% (+/-0.00%)
bm_float.py                   5872.71 ->    5910.11 :     +37.40 =  +0.637% (+/-0.01%)
bm_hexiom.py                    48.77 ->      48.85 :      +0.08 =  +0.164% (+/-0.00%)
bm_nqueens.py                 4433.80 ->    4421.59 :     -12.21 =  -0.275% (+/-0.00%)
bm_pidigits.py                 640.95 ->     644.49 :      +3.54 =  +0.552% (+/-0.25%)
core_import_mpy_multi.py       759.37 ->     763.36 :      +3.99 =  +0.525% (+/-0.00%)
core_import_mpy_single.py      105.41 ->     106.25 :      +0.84 =  +0.797% (+/-0.01%)
core_qstr.py                   206.44 ->     212.31 :      +5.87 =  +2.843% (+/-0.00%)
core_yield_from.py             353.96 ->     354.05 :      +0.09 =  +0.025% (+/-0.00%)
misc_aes.py                    410.03 ->     404.15 :      -5.88 =  -1.434% (+/-0.00%)
misc_mandel.py                3387.86 ->    3397.47 :      +9.61 =  +0.284% (+/-0.00%)
misc_pystone.py               2419.06 ->    2446.53 :     +27.47 =  +1.136% (+/-0.00%)
misc_raytrace.py               368.50 ->     369.21 :      +0.71 =  +0.193% (+/-0.00%)
```
The `core_qstr.py` test is intended to really stress test the `qstr_find_strn()` function, so I'd say these results give an accurate picture.

Those tests are run with lots of frozen code on the PYBv1.0, and frozen qstrs are not sorted.  So might be better to try again without frozen code.

dpgeorge commented on py/qstr.c:
```diff
@@ -179,13 +228,45 @@ STATIC qstr qstr_add(mp_uint_t hash, mp_uint_t len, const char *q_ptr) {
+                } else {
+                    // avoid a rare (hash,len) collisions
+                    while (MP_UNLIKELY(
+                        pool->lengths[mid] != str_len ||
```
I don't think you need this check on length.  If the code gets here then `cmp == 0`, which can only happen if the lengths are equal (and also the hashes are equal).

dpgeorge replied:
And wouldn't it be simpler (smaller code, probably no effect on performance) to just break here and drop down into the sequential search?

amirgon replied:
You are right - this can be simplified.  
Done in 3bc257c4b24d5d2e4c58c13a91a31f1a58306bdd.

amirgon commented:
> I did performance tests for sorting on strings and sorting on hashes (on PYBv1.0).

@dpgeorge did you try performance tests on PYB with different values of `MP_QSTR_SEARCH_THRESHOLD`?

amirgon committed:
py/qstr: Optimize hash sort.

In case cmp == 0 there could be a hash collision. To save program size
Instead of checking for hash collisions, simply revert to linear search
which also compares strings.

Signed-off-by: Amir Gonnen <amirgonnen@gmail.com>

amirgon commented:
> Those tests are run with lots of frozen code on the PYBv1.0, and frozen qstrs are not sorted. So might be better to try again without frozen code.

@dpgeorge Are you sure frozen qstrs are not sorted?? I thought I took care of that.

https://github.com/micropython/micropython/blob/660c81085791cf011279a9e4669413f483df44ad/tools/mpy-tool.py#L1276

https://github.com/micropython/micropython/blob/660c81085791cf011279a9e4669413f483df44ad/tools/mpy-tool.py#L1360

AmirHmZz commented:
@amirgon Thanks for awesome work done here!
Any updates on this?

dpgeorge commented:
I did some more testing on this.  I see that in the current state this PR does not sort the first set of qstrs, the static ones, and the `__mul__` etc ones that need to come first so their value fits in 1 byte.  So that means the searching is not improved for this first set (first pool) of qstrs.  This seems like a wasted opportunity, especially since this first set will always be searched if the qstr being search for does not exist yet.

I did a test to see how much performance improves if this first set of qstrs are sorted.  The improvement is quite substantial (run on PYBv1.0 with about 1400 built-in qstrs and 270 frozen qstrs).

When sorting by string value, sorting also the first pool increases performance by:
```
diff of scores (higher is better)
N=100 M=100           p-tail-sort-str -> p-all-sort-str     diff      diff% (error%)
bm_chaos.py                    365.25 ->     367.44 :      +2.19 =  +0.600% (+/-0.01%)
bm_fannkuch.py                  76.88 ->      77.78 :      +0.90 =  +1.171% (+/-0.01%)
bm_fft.py                     2526.83 ->    2553.02 :     +26.19 =  +1.036% (+/-0.00%)
bm_float.py                   5887.79 ->    5925.10 :     +37.31 =  +0.634% (+/-0.02%)
bm_hexiom.py                    48.29 ->      48.32 :      +0.03 =  +0.062% (+/-0.00%)
bm_nqueens.py                 4395.08 ->    4385.09 :      -9.99 =  -0.227% (+/-0.00%)
bm_pidigits.py                 629.43 ->     636.36 :      +6.93 =  +1.101% (+/-0.35%)
core_import_mpy_multi.py       584.97 ->     718.14 :    +133.17 = +22.765% (+/-0.00%)
core_import_mpy_single.py       85.38 ->     114.16 :     +28.78 = +33.708% (+/-0.01%)
core_qstr.py                   205.08 ->     307.38 :    +102.30 = +49.883% (+/-0.01%)
core_yield_from.py             357.32 ->     355.98 :      -1.34 =  -0.375% (+/-0.00%)
misc_aes.py                    411.99 ->     417.28 :      +5.29 =  +1.284% (+/-0.01%)
misc_mandel.py                3353.10 ->    3425.33 :     +72.23 =  +2.154% (+/-0.00%)
misc_pystone.py               2254.72 ->    2352.82 :     +98.10 =  +4.351% (+/-0.01%)
misc_raytrace.py               369.16 ->     375.49 :      +6.33 =  +1.715% (+/-0.01%)
viper_call0.py                 576.78 ->     576.68 :      -0.10 =  -0.017% (+/-0.00%)
viper_call1a.py                550.32 ->     550.32 :      +0.00 =  +0.000% (+/-0.00%)
viper_call1b.py                438.29 ->     436.01 :      -2.28 =  -0.520% (+/-0.07%)
viper_call1c.py                442.90 ->     441.27 :      -1.63 =  -0.368% (+/-0.23%)
viper_call2a.py                536.26 ->     536.26 :      +0.00 =  +0.000% (+/-0.00%)
viper_call2b.py                378.94 ->     377.36 :      -1.58 =  -0.417% (+/-0.15%)
```

When sorting by hash, the improvement when sorting the first pool is:
```
diff of scores (higher is better)
N=100 M=100          p-tail-sort-hash -> p-all-sort-hash         diff      diff% (error%)
bm_chaos.py                    366.20 ->     364.72 :      -1.48 =  -0.404% (+/-0.01%)
bm_fannkuch.py                  77.04 ->      77.00 :      -0.04 =  -0.052% (+/-0.01%)
bm_fft.py                     2520.72 ->    2545.80 :     +25.08 =  +0.995% (+/-0.00%)
bm_float.py                   5905.63 ->    5850.64 :     -54.99 =  -0.931% (+/-0.00%)
bm_hexiom.py                    48.31 ->      47.97 :      -0.34 =  -0.704% (+/-0.00%)
bm_nqueens.py                 4400.15 ->    4351.58 :     -48.57 =  -1.104% (+/-0.00%)
bm_pidigits.py                 630.84 ->     632.89 :      +2.05 =  +0.325% (+/-0.34%)
core_import_mpy_multi.py       692.19 ->     717.92 :     +25.73 =  +3.717% (+/-0.00%)
core_import_mpy_single.py      104.98 ->     116.19 :     +11.21 = +10.678% (+/-0.01%)
core_qstr.py                   210.76 ->     332.36 :    +121.60 = +57.696% (+/-0.00%)
core_yield_from.py             356.26 ->     356.72 :      +0.46 =  +0.129% (+/-0.00%)
misc_aes.py                    411.77 ->     416.57 :      +4.80 =  +1.166% (+/-0.00%)
misc_mandel.py                3355.21 ->    3358.47 :      +3.26 =  +0.097% (+/-0.01%)
misc_pystone.py               2269.62 ->    2349.21 :     +79.59 =  +3.507% (+/-0.01%)
misc_raytrace.py               372.72 ->     365.83 :      -6.89 =  -1.849% (+/-0.01%)
viper_call0.py                 576.67 ->     576.67 :      +0.00 =  +0.000% (+/-0.00%)
viper_call1a.py                550.32 ->     550.32 :      +0.00 =  +0.000% (+/-0.00%)
viper_call1b.py                438.20 ->     436.46 :      -1.74 =  -0.397% (+/-0.21%)
viper_call1c.py                442.81 ->     442.00 :      -0.81 =  -0.183% (+/-0.10%)
viper_call2a.py                536.26 ->     536.26 :      +0.00 =  +0.000% (+/-0.00%)
viper_call2b.py                378.85 ->     377.37 :      -1.48 =  -0.391% (+/-0.15%)
```

Then comparing string sort to hash sort, where everything is sorted (there is a small gain to sorting by hash):
```
diff of scores (higher is better)
N=100 M=100          p-all-sort-str -> p-all-sort-hash         diff      diff% (error%)
bm_chaos.py                    367.44 ->     364.72 :      -2.72 =  -0.740% (+/-0.01%)
bm_fannkuch.py                  77.78 ->      77.00 :      -0.78 =  -1.003% (+/-0.01%)
bm_fft.py                     2553.02 ->    2545.80 :      -7.22 =  -0.283% (+/-0.00%)
bm_float.py                   5925.10 ->    5850.64 :     -74.46 =  -1.257% (+/-0.02%)
bm_hexiom.py                    48.32 ->      47.97 :      -0.35 =  -0.724% (+/-0.00%)
bm_nqueens.py                 4385.09 ->    4351.58 :     -33.51 =  -0.764% (+/-0.00%)
bm_pidigits.py                 636.36 ->     632.89 :      -3.47 =  -0.545% (+/-0.34%)
core_import_mpy_multi.py       718.14 ->     717.92 :      -0.22 =  -0.031% (+/-0.00%)
core_import_mpy_single.py      114.16 ->     116.19 :      +2.03 =  +1.778% (+/-0.01%)
core_qstr.py                   307.38 ->     332.36 :     +24.98 =  +8.127% (+/-0.00%)
core_yield_from.py             355.98 ->     356.72 :      +0.74 =  +0.208% (+/-0.00%)
misc_aes.py                    417.28 ->     416.57 :      -0.71 =  -0.170% (+/-0.01%)
misc_mandel.py                3425.33 ->    3358.47 :     -66.86 =  -1.952% (+/-0.01%)
misc_pystone.py               2352.82 ->    2349.21 :      -3.61 =  -0.153% (+/-0.01%)
misc_raytrace.py               375.49 ->     365.83 :      -9.66 =  -2.573% (+/-0.01%)
viper_call0.py                 576.68 ->     576.67 :      -0.01 =  -0.002% (+/-0.00%)
viper_call1a.py                550.32 ->     550.32 :      +0.00 =  +0.000% (+/-0.00%)
viper_call1b.py                436.01 ->     436.46 :      +0.45 =  +0.103% (+/-0.23%)
viper_call1c.py                441.27 ->     442.00 :      +0.73 =  +0.165% (+/-0.25%)
viper_call2a.py                536.26 ->     536.26 :      +0.00 =  +0.000% (+/-0.00%)
viper_call2b.py                377.36 ->     377.37 :      +0.01 =  +0.003% (+/-0.21%)
```

Finally, to compare baseline to sorting everything using string:
```
diff of scores (higher is better)
N=100 M=100            p-baseline -> p-all-sort-str         diff      diff% (error%)
bm_chaos.py                    362.49 ->     367.44 :      +4.95 =  +1.366% (+/-0.01%)
bm_fannkuch.py                  76.92 ->      77.78 :      +0.86 =  +1.118% (+/-0.01%)
bm_fft.py                     2520.39 ->    2553.02 :     +32.63 =  +1.295% (+/-0.00%)
bm_float.py                   5889.37 ->    5925.10 :     +35.73 =  +0.607% (+/-0.02%)
bm_hexiom.py                    47.05 ->      48.32 :      +1.27 =  +2.699% (+/-0.00%)
bm_nqueens.py                 4409.34 ->    4385.09 :     -24.25 =  -0.550% (+/-0.00%)
bm_pidigits.py                 632.59 ->     636.36 :      +3.77 =  +0.596% (+/-0.46%)
core_import_mpy_multi.py       470.16 ->     718.14 :    +247.98 = +52.744% (+/-0.00%)
core_import_mpy_single.py       58.06 ->     114.16 :     +56.10 = +96.624% (+/-0.01%)
core_qstr.py                    63.49 ->     307.38 :    +243.89 = +384.139% (+/-0.02%)
core_yield_from.py             356.27 ->     355.98 :      -0.29 =  -0.081% (+/-0.00%)
misc_aes.py                    411.47 ->     417.28 :      +5.81 =  +1.412% (+/-0.01%)
misc_mandel.py                3376.85 ->    3425.33 :     +48.48 =  +1.436% (+/-0.00%)
misc_pystone.py               2263.46 ->    2352.82 :     +89.36 =  +3.948% (+/-0.01%)
misc_raytrace.py               369.15 ->     375.49 :      +6.34 =  +1.717% (+/-0.01%)
viper_call0.py                 576.76 ->     576.68 :      -0.08 =  -0.014% (+/-0.01%)
viper_call1a.py                550.32 ->     550.32 :      +0.00 =  +0.000% (+/-0.00%)
viper_call1b.py                438.29 ->     436.01 :      -2.28 =  -0.520% (+/-0.07%)
viper_call1c.py                442.90 ->     441.27 :      -1.63 =  -0.368% (+/-0.23%)
viper_call2a.py                536.26 ->     536.26 :      +0.00 =  +0.000% (+/-0.00%)
viper_call2b.py                378.93 ->     377.36 :      -1.57 =  -0.414% (+/-0.15%)
```

And baseline versus sorting everything by hash:
```
diff of scores (higher is better)
N=100 M=100            p-baseline -> p-all-sort-hash         diff      diff% (error%)
bm_chaos.py                    362.49 ->     364.72 :      +2.23 =  +0.615% (+/-0.01%)
bm_fannkuch.py                  76.92 ->      77.00 :      +0.08 =  +0.104% (+/-0.01%)
bm_fft.py                     2520.39 ->    2545.80 :     +25.41 =  +1.008% (+/-0.00%)
bm_float.py                   5889.37 ->    5850.64 :     -38.73 =  -0.658% (+/-0.00%)
bm_hexiom.py                    47.05 ->      47.97 :      +0.92 =  +1.955% (+/-0.00%)
bm_nqueens.py                 4409.34 ->    4351.58 :     -57.76 =  -1.310% (+/-0.00%)
bm_pidigits.py                 632.59 ->     632.89 :      +0.30 =  +0.047% (+/-0.42%)
core_import_mpy_multi.py       470.16 ->     717.92 :    +247.76 = +52.697% (+/-0.00%)
core_import_mpy_single.py       58.06 ->     116.19 :     +58.13 = +100.121% (+/-0.01%)
core_qstr.py                    63.49 ->     332.36 :    +268.87 = +423.484% (+/-0.02%)
core_yield_from.py             356.27 ->     356.72 :      +0.45 =  +0.126% (+/-0.00%)
misc_aes.py                    411.47 ->     416.57 :      +5.10 =  +1.239% (+/-0.01%)
misc_mandel.py                3376.85 ->    3358.47 :     -18.38 =  -0.544% (+/-0.01%)
misc_pystone.py               2263.46 ->    2349.21 :     +85.75 =  +3.788% (+/-0.00%)
misc_raytrace.py               369.15 ->     365.83 :      -3.32 =  -0.899% (+/-0.01%)
viper_call0.py                 576.76 ->     576.67 :      -0.09 =  -0.016% (+/-0.01%)
viper_call1a.py                550.32 ->     550.32 :      +0.00 =  +0.000% (+/-0.00%)
viper_call1b.py                438.29 ->     436.46 :      -1.83 =  -0.418% (+/-0.21%)
viper_call1c.py                442.90 ->     442.00 :      -0.90 =  -0.203% (+/-0.10%)
viper_call2a.py                536.26 ->     536.26 :      +0.00 =  +0.000% (+/-0.00%)
viper_call2b.py                378.93 ->     377.37 :      -1.56 =  -0.412% (+/-0.15%)
```

-------------

Summary:
- This scheme only sorts qstrs that are part of the firmware, it does not sort qstrs that are added dynamically at runtime.  But still it gives a great improvement to performance.
- It's definitely worth sorting all the qstrs, including the first set (the pre-defined ones, and the ones that need to fit in 1 byte).
- Sorting via hash is slightly faster than sorting via string value.
- Sorting via hash will change the order of tab completion, but IMO that is OK because tab completion is not anyway sorted when it's completing a user defined object that has qstrs added at runtime.
- Sorting the first set of qstrs via hash will move `MP_QSTR_` away from the second position.  I'm not 100% sure this is OK, will need to be checked if nothing relies on that qstr having value 1.
- Sorting the first set of qstrs via hash will need builds that use 1-byte qstr hashes to use the most-significant-byte of the hash, so that the order of the pre-defined qstrs is the same regardless of whether 1-byte or 2-byte qstr hashes are used.

A final decision needs to be made whether to sort by hash or string value.  Sorting by hash has some complexity to it.

dpgeorge commented:
Here is the code size diff.  For baseline -> this PR with string sorting:
```
   bare-arm:  +104 +0.180% 
minimal x86:  +309 +0.187% [incl +104(data)]
   unix x64:  +336 +0.064% [incl +96(data)]
unix nanbox:  +228 +0.049% [incl +32(data)]
      stm32:  +104 +0.026% PYBV10
     cc3200:   +96 +0.052% 
    esp8266:  +176 +0.025% GENERIC
      esp32:  +132 +0.009% GENERIC[incl +32(data)]
        nrf:   +88 +0.048% pca10040
        rp2:   +56 +0.011% PICO
       samd:  +104 +0.074% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

For baseline -> this PR with hash sorting:
```
   bare-arm:   +96 +0.166% 
minimal x86:  +296 +0.180% [incl +72(data)]
   unix x64:  +304 +0.058% [incl +64(data)]
unix nanbox:  +208 +0.045% [incl +32(data)]
      stm32:  +100 +0.025% PYBV10
     cc3200:   +96 +0.052% 
    esp8266:  +192 +0.028% GENERIC
      esp32:  +124 +0.008% GENERIC[incl +32(data)]
        nrf:   +76 +0.041% pca10040
        rp2:   +56 +0.011% PICO
       samd:   +72 +0.051% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

The difference between the above, ie string -> hash sorting:
```
   bare-arm:    -8 -0.014% 
minimal x86:   -13 -0.008% [incl -32(data)]
   unix x64:   -32 -0.006% [incl -32(data)]
unix nanbox:   -20 -0.004% 
      stm32:    -4 -0.001% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:   +16 +0.002% GENERIC
      esp32:    -8 -0.001% GENERIC
        nrf:   -12 -0.007% pca10040
        rp2:    +0 +0.000% PICO
       samd:   -32 -0.023% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

Sorting using hash is only slightly smaller in code size.

The code size increase is quite large... but I think there might be a few opportunities to reduce that, eg shrink the size of the `qstr_pool_t` struct.

stinos commented:
Seeing the perf numbers it's definitely worth to have this, the hash version I'd asy, and indeed a build option seems ok and default enabled on ports with enough RAM?

andrewleech commented:
It would probably make sense to link such a setting to 
MICROPY_CONFIG_ROM_LEVEL_MINIMUM or CORE_FEATURES.

amirgon committed:
py/qstr: Fix hash collision corner case.

This fixes a corner case where both hash and lengths of two qstrs are
equal. In such case, if the binary search lower bound is the second
instance of the same hash, the search will miss the first instance.
To overcome this, when lower bound hash equals the requested hash, we move
the lower bound backwards to cover all previous hashes which are also equal
to the requested hash.
Also removed length comparison since it's no longer needed.

Related: https://github.com/lvgl/lv_binding_micropython/issues/224

Signed-off-by: Amir Gonnen <amirgonnen@gmail.com>

amirgon committed:
py/qstr: Remove MP_QSTR_SEARCH_THRESHOLD.

MP_QSTR_SEARCH_THRESHOLD doesn't provide improvement with sorted hashes.
Remove MP_QSTR_SEARCH_THRESHOLD and simlify the binary search.

Sequential search is still needed in case of hash collisions, or
unsorted pools.

Signed-off-by: Amir Gonnen <amirgonnen@gmail.com>

amirgon committed:
py/qstr: Sort the first QSTR pool.

Signed-off-by: Amir Gonnen <amirgonnen@gmail.com>

amirgon commented:
@dpgeorge 

> It's definitely worth sorting all the qstrs, including the first set (the pre-defined ones, and the ones that need to fit in 1 byte).

I agree.
Done on 04ba17bff82c8d0083b42ea8eed2c9b575a1902a.

> Sorting the first set of qstrs via hash will move MP_QSTR_ away from the second position. I'm not 100% sure this is OK, will need to be checked if nothing relies on that qstr having value 1.

I'm not sure how to check this really.

> Sorting the first set of qstrs via hash will need builds that use 1-byte qstr hashes to use the most-significant-byte of the hash, so that the order of the pre-defined qstrs is the same regardless of whether 1-byte or 2-byte qstr hashes are used.

Once the first set is sorted by hash, why do we care if the order is different between 1-byte and 2-byte qstr hashes?

> The code size increase is quite large... but I think there might be a few opportunities to reduce that, eg shrink the size of the qstr_pool_t struct.

I don't understand why we see a large code increase.  
The code change is rather small, the increase per pool is small and there are usually small number of pools. We could even keep `qstr_pool_t` the same size if we encoded `sorted` in a bit of one of the other fields.  

But overall, where does the significant code increase come from?  
Maybe the compiler is unrolling the loop? But when code size is important we would compile with `-Os` and prevent loop unrolling.   
Did you measure the code increase when compiling with `-Os`?

jimmo commented:
Thanks @amirgon for your work on this and for bringing it to our attention in the first place -- this has been merged via #12678 which contains essentially the same change to qstr.c (but via a fairly circuitous route, see #10758, where we investigated a bunch of other things). I think it's still worth investigating splitting the static pool, and I'd love to eventually find a way to remove hashes without the performance hit.

dpgeorge commented:
> One thing we can consider in order to decrease memory usage is to remove qstr lengths array entirely.

I had a brief look at doing this and I think it would slow a lot of operations down, because there are many places in the code that query the qstr/str lengths.  Eg all uses of `GET_STR_DATA_LEN()`, that would need to use `strlen()` to compute the qstr length each time it was called.

gitcnd commented:
What are all the qstr's for?  How many are used in a normal application?  Eradicating as many as possible might be a more sensible option?  Wouldn't making the caller turn whatever they want into an ENUM would make more sense?
